internal/pkg/status/legacy_status.go

```diff
@@ -0,0 +1,238 @@
+	}
+
+	for _, c := range conditions {
+		switch clientu.GetStringField(c, "type", "") {
+		case "Progressing": //appsv1.DeploymentProgressing:
+			// https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/pkg/controller/deployment/progress.go#L52
+			status := clientu.GetStringField(c, "status", "")
+			reason := clientu.GetStringField(c, "reason", "")
+			if status != "True" || reason != "NewReplicaSetAvailable" {
+				progress = false
+			}
+		case "Available": //appsv1.DeploymentAvailable:
+			status := clientu.GetStringField(c, "status", "")
+			if status == "False" {
+				available = false
+			}
+		}
+	}
+
+	return progress && available, nil
+}
+
+// Replicaset
+func replicasetReady(u *unstructured.Unstructured) (bool, error) {
+	failure := false
+	conditions := clientu.GetConditions(u.UnstructuredContent())
+
+	for _, c := range conditions {
+		switch clientu.GetStringField(c, "type", "") {
+		// https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/pkg/controller/replicaset/replica_set_utils.go
+		case "ReplicaFailure": //appsv1.ReplicaSetReplicaFailure
+			status := clientu.GetStringField(c, "status", "")
+			if status == "True" {
+				failure = true
+				break
+			}
+		}
+	}
+
+	c1, err := compareIntFields(u, []string{"status", "replicas"}, []string{"status", "readyReplicas"}, equalInt)
+	if err != nil {
+		return c1, err
+	}
+	c2, err := compareIntFields(u, []string{"status", "replicas"}, []string{"status", "availableReplicas"}, equalInt)
+	if err != nil {
+		return c2, err
+	}
+
+	return !failure && c1 && c2, nil
+}
+
+// Daemonset
+func daemonsetReady(u *unstructured.Unstructured) (bool, error) {
+	c1, err := compareIntFields(u, []string{"status", "desiredNumberScheduled"}, []string{"status", "numberAvailable"}, equalInt)
+	if err != nil {
+		return c1, err
+	}
+	c2, err := compareIntFields(u, []string{"status", "desiredNumberScheduled"}, []string{"status", "numberReady"}, equalInt)
+	if err != nil {
+		return c2, err
+	}
+
+	return c1 && c2, nil
+}
+
+// PVC
+func pvcReady(u *unstructured.Unstructured) (bool, error) {
+	val, found, err := unstructured.NestedString(u.UnstructuredContent(), "status", "phase")
+	if err != nil {
+		return false, err
+	}
+	if !found {
+		return false, fmt.Errorf(".status.phase not found")
+	}
+	return val == "Bound", nil // corev1.ClaimBound
+}
+
+// Pod
+func podReady(u *unstructured.Unstructured) (bool, error) {
+	conditions := clientu.GetConditions(u.UnstructuredContent())
+
+	for _, c := range conditions {
+		if clientu.GetStringField(c, "type", "") == "Ready" && (clientu.GetStringField(c, "status", "") == "True" || clientu.GetStringField(c, "reason", "") == "PodCompleted") {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// PodDisruptionBudget
+func pdbReady(u *unstructured.Unstructured) (bool, error) {
+	return compareIntFields(u, []string{"status", "currentHealthy"}, []string{"status", "desiredHealthy"}, geInt)
```

Comment:
When getting back the YAML immediately after applying, with this logic, pdb would return true even though it's not actually ready:
```
➜  kubectl apply -f test/configs-all/pdb.yaml -n test-local-deploy-all -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"policy/v1beta1","kind":"PodDisruptionBudget","metadata":{"annotations":{},"name":"test-app-pdb","namespace":"test-local-deploy-all"},"spec":{"minAvailable":2,"selector":{"matchLabels":{"app":"test-app"}}}}
  creationTimestamp: 2019-06-11T15:14:19Z
  generation: 1
  name: test-app-pdb
  namespace: test-local-deploy-all
  resourceVersion: "6036557"
  selfLink: /apis/policy/v1beta1/namespaces/test-local-deploy-all/poddisruptionbudgets/test-app-pdb
  uid: 958e7b84-8c5b-11e9-8840-42010a8e00dc
spec:
  minAvailable: 2
  selector:
    matchLabels:
      app: test-app
status:
  currentHealthy: 0
  desiredHealthy: 0
  disruptionsAllowed: 0
  expectedPods: 0
```

Reply:
If there is are more pods than `minAvailable`, the pdb will never return ready:
```
➜  kubectl get pdb test-app-pdb -n test-local-deploy-all -o yaml           
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"policy/v1beta1","kind":"PodDisruptionBudget","metadata":{"annotations":{},"name":"test-app-pdb","namespace":"test-local-deploy-all"},"spec":{"minAvailable":2,"selector":{"matchLabels":{"app":"test-app"}}}}
  creationTimestamp: 2019-06-11T15:14:19Z
  generation: 1
  name: test-app-pdb
  namespace: test-local-deploy-all
  resourceVersion: "6036558"
  selfLink: /apis/policy/v1beta1/namespaces/test-local-deploy-all/poddisruptionbudgets/test-app-pdb
  uid: 958e7b84-8c5b-11e9-8840-42010a8e00dc
spec:
  minAvailable: 2
  selector:
    matchLabels:
      app: test-app
status:
  currentHealthy: 11
  desiredHealthy: 2
  disruptionsAllowed: 9
  expectedPods: 11
  observedGeneration: 1
```

Reply:
Can you tell me how you came about this logic?

Reply:
Another case: There are no pods that match the label:
```
➜  kubectl get pdb test-app-pdb -n test-local-deploy-all -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"policy/v1beta1","kind":"PodDisruptionBudget","metadata":{"annotations":{},"name":"test-app-pdb","namespace":"test-local-deploy-all"},"spec":{"minAvailable":2,"selector":{"matchLabels":{"app":"test-app-deployment"}}}}
  creationTimestamp: 2019-06-11T15:11:58Z
  generation: 1
  name: test-app-pdb
  namespace: test-local-deploy-all
  resourceVersion: "6036053"
  selfLink: /apis/policy/v1beta1/namespaces/test-local-deploy-all/poddisruptionbudgets/test-app-pdb
  uid: 41d83cf7-8c5b-11e9-8840-42010a8e00dc
spec:
  minAvailable: 2
  selector:
    matchLabels:
      app: test-app-deployment
status:
  currentHealthy: 0
  desiredHealthy: 2
  disruptionsAllowed: 0
  expectedPods: 0
  observedGeneration: 1
```

Reply:
Another case: `minAvailable` is 0:
```
➜  kubectl get pdb test-app-pdb -n test-local-deploy-all -o yaml   
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"policy/v1beta1","kind":"PodDisruptionBudget","metadata":{"annotations":{},"name":"test-app-pdb","namespace":"test-local-deploy-all"},"spec":{"minAvailable":0,"selector":{"matchLabels":{"app":"test-app"}}}}
  creationTimestamp: 2019-06-11T15:21:07Z
  generation: 1
  name: test-app-pdb
  namespace: test-local-deploy-all
  resourceVersion: "6037971"
  selfLink: /apis/policy/v1beta1/namespaces/test-local-deploy-all/poddisruptionbudgets/test-app-pdb
  uid: 88fa639a-8c5c-11e9-8840-42010a8e00dc
spec:
  minAvailable: 0
  selector:
    matchLabels:
      app: test-app
status:
  currentHealthy: 11
  desiredHealthy: 0
  disruptionsAllowed: 11
  expectedPods: 11
  observedGeneration: 1
```

Reply:
This would be Ready since minAvailable is met.

Reply:
Apologies. I didn't notice the conditional was "geInt" and assumed it was comparing for equality